tests/ttnn/unit_tests/operations/test_all_gather_matmul.py

```diff
@@ -136,11 +140,33 @@ def run_all_gather_matmul_on_t3000_impl(
             compute_kernel_config=compute_kernel_config,
         )
 
-        logger.info(f"Done iteration {i}")
+    if enable_trace:
+        ##### Capture Trace #####
+
+        # Compile the op
+        tt_all_gather_out_tensor, tt_matmul_output, tt_datacopy_out_tensor = run_op()
+        logger.info(f"Done compiling Op")
 
-        # Synchronize the devices
-        for d in devices:
-            ttnn.synchronize_device(d)
+        # Capture the trace
+        trace_id = ttnn.begin_trace_capture(t3k_mesh_device)
+        for i in range(num_iters):
+            tt_all_gather_out_tensor, tt_matmul_output, tt_datacopy_out_tensor = run_op()
+        ttnn.end_trace_capture(t3k_mesh_device, trace_id)
+
+        logger.info(f"Done capturing trace")
+
+        # Execute trace
+        ttnn.execute_trace(t3k_mesh_device, trace_id, blocking=False)
+        logger.info(f"Done executing trace")
+    else:
+        for i in range(num_iters):
+            tt_all_gather_out_tensor, tt_matmul_output, tt_datacopy_out_tensor = run_op()
+
+            # Synchronize the devices
+            for d in devices:
+                ttnn.synchronize_device(d)
```

Comment:
Why are devices syncrhonized only when trace is disabled? Does ttnn.begin_trace_capture force synchronization?

Reply:
will clarify this with the metal team

Reply:
ttnn begin trace doesn't capture (synchronize) events. Trying to synchronize should result in an error (I saw similar behaviour and @tt-asaigal confirmed with me).

Reply:
Wouldn't the two code paths have different behaviour if one synchronizes and one doesn't?

Reply:
In what sense? The execution schedule will naturally change when switching between trace and non-trace mode. But that being said, to be "more consistent", the `execute_trace` call should probably just set `blocking=True`, which would be equivalent to calling synchronize at the end. Is that what you were getting at?

Reply:
In the sense that the schedules should be as similar as possible.

Yes, that's what I was getting at.

Reply:
Currently, using blocking=True for all gather with trace mode enabled hangs. See issue with minimal repro here: #12481 .